### PR TITLE
tweak(core): use Deferred instad of Ref

### PR DESCRIPTION
**Pull request by i10416**
I'm not sure if it was intentional to use `Ref` instead of `Deferred`, but It seems simpler to use `Deferred` as status is set exactly once.

### CODE REVIEW

File: core/src/main/scala/org/http4s/grpc/ServerGrpc.scala [hunk 1]
```diff
@@ -40,10 +40,10 @@ object ServerGrpc {
           .flatMap(codecs.Messages.encodeSingle(encode)(_))
           .through(timeoutStream(_)(timeout.map(_.duration)))
           .onFinalizeCase {
```
Comment (danicheg) [inline core/src/main/scala/org/http4s/grpc/ServerGrpc.scala:42]: I think we should use `onFinalizeCaseWeak` here. @hamnis wdyt?

**Review by danicheg — COMMENTED** [review]
Generally, using `Deferred` is quite natural in the context of Trailers. However, since `Deferred` is related to synchronization, we should be extra careful about filling it with a value, as we could unintentionally run into an async block. See the http4s API:
```scala
/** The trailer headers, as specified in Section 3.6.1 of RFC 2616. The resulting
  * F might not complete until the entire body has been consumed.
  */
def trailerHeaders(implicit F: Applicative[F]): F[Headers]
```

File: core/src/main/scala/org/http4s/grpc/ServerGrpc.scala (hunk 1 above)
Comment (hamnis) [reply core/src/main/scala/org/http4s/grpc/ServerGrpc.scala:42]: since we don't need a new scope here, i think that is reasonable.

Comment (i10416) [reply core/src/main/scala/org/http4s/grpc/ServerGrpc.scala]: Thank you for your advice. I applied the suggestion.

**Comment by danicheg** [comment]
Thanks!

**Pull request merged**